riscv-c-api.md

```diff
@@ -114,7 +129,96 @@ For example:
+| `long _rv_sha256sig0(long rs1);`                              | `sha256sig0`  | Sigma0 function for SHA2-256.             | Zknh, Zkn, Zk (RV32,RV64)     |
+| `long _rv_sha256sig1(long rs1);`                              | `sha256sig1`  | Sigma1 function for SHA2-256.             | Zknh, Zkn, Zk (RV32,RV64)     |
+| `long _rv_sha256sum0(long rs1);`                              | `sha256sum0`  | Sum0 function for SHA2-256.               | Zknh, Zkn, Zk (RV32,RV64)     |
+| `long _rv_sha256sum1(long rs1);`                              | `sha256sum1`  | Sum1 function for SHA2-256.               | Zknh, Zkn, Zk (RV32,RV64)     |
+| `int32_t _rv32_sha512sig0h(int32_t rs1, int32_t rs2);`        | `sha512sig0h` | Sigma0 high half for SHA2-512.            | Zknh, Zkn, Zk (RV32)          |
+| `int32_t _rv32_sha512sig0l(int32_t rs1, int32_t rs2);`        | `sha512sig0l` | Sigma0 low half for SHA2-512.             | Zknh, Zkn, Zk (RV32)          |
+| `int32_t _rv32_sha512sig1h(int32_t rs1, int32_t rs2);`        | `sha512sig1h` | Sigma1 high half for SHA2-512.            | Zknh, Zkn, Zk (RV32)          |
+| `int32_t _rv32_sha512sig1l(int32_t rs1, int32_t rs2);`        | `sha512sig1l` | Sigma1 low half for SHA2-512.             | Zknh, Zkn, Zk (RV32)          |
+| `int32_t _rv32_sha512sum0r(int32_t rs1, int32_t rs2);`        | `sha512sum0r` | Sum0 function for SHA2-512.               | Zknh, Zkn, Zk (RV32)          |
+| `int32_t _rv32_sha512sum1r(int32_t rs1, int32_t rs2);`        | `sha512sum1r` | Sum1 function for SHA2-512.               | Zknh, Zkn, Zk (RV32)          |
+| `int64_t _rv64_sha512sig0(int64_t rs1);`                      | `sha512sig0`  | Sigma0 function for SHA2-512.             | Zknh, Zkn, Zk (RV64 [RV32])   |
+| `int64_t _rv64_sha512sig1(int64_t rs1);`                      | `sha512sig1`  | Sigma1 function for SHA2-512.             | Zknh, Zkn, Zk (RV64 [RV32])   |
+| `int64_t _rv64_sha512sum0(int64_t rs1);`                      | `sha512sum0`  | Sum0 function for SHA2-512.               | Zknh, Zkn, Zk (RV64 [RV32])   |
+| `int64_t _rv64_sha512sum1(int64_t rs1);`                      | `sha512sum1`  | Sum1 function for SHA2-512.               | Zknh, Zkn, Zk (RV64 [RV32])   |
+| `long _rv_sm3p0(long rs1);`                                   | `sm3p0`       | P0 function for SM3 hash.                 | Zksh, Zks (RV32,RV64)         |
+| `long _rv_sm3p1(long rs1);`                                   | `sm3p1`       | P1 function for SM3 hash.                 | Zksh, Zks (RV32,RV64)         |
+| `long _rv_sm4ed(int32_t rs1, int32_t rs2, int bs);`           | `sm4ed`       | Accelerate SM4 cipher encrypt/decrypt.    | Zksed, Zks (RV32,RV64)        |
+| `long _rv_sm4ks(int32_t rs1, int32_t rs2, int bs);`           | `sm4ks`       | Accelerate SM4 cipher key schedule.       | Zksed, Zks (RV32,RV64)        |
```

Comment:
Is the return type of long important here or could it be int32_t? I know the result returns a sign extended value on RV64. _rv32_ror doesn't return long for 64-bits. Or is that because _rv32_ror came from bitmanip and if you defined it from scratch it would return long?

Reply:
It was defined this way in bitmanip's `rvintrin.h`. I agree that 32-bit ror and rol could be defined with "long" for consistency, as it is available in both RV32 and RV64. On the other hand I see no great danger in keeping it this way.

Reply:
Instructions that are guaranteed to return a sign-extended 32-bit result should definitely return `int32_t` or equivalent.  This tightens up the API a little bit, and it'll probably obviate some stray `sext.w` instructions in some cases under some compilers.